sys/pm_callouts.c

```diff
@@ -408,6 +408,191 @@ void redir(portmaster_packet_info* packetInfo, portmaster_packet_info* redirInfo
     return;
 }
 
+void send_tcp_rst(portmaster_packet_info* packetInfo, void* originalPacket, ULONG originalPacketLength) {
+    // Only TCP is supported
+    if(packetInfo->protocol != 6) {
+        return;
+    }
+
+    if(packetInfo->ipV6) {
+        // Initialize header for the original packet with SYN flag
+        ULONG originalIPHeaderLength = calc_ipv6_header_size(originalPacket, originalPacketLength, NULL);
+        PIPV6_HEADER originalIPHeader = (PIPV6_HEADER) originalPacket;
+        PTCP_HEADER originalTCPHeader = (PTCP_HEADER) ((UINT8*)originalPacket + originalIPHeaderLength);
+
+        // Initialize variables
+        PNET_BUFFER_LIST injectNBL;
+        NTSTATUS status;
+        UINT16 packetLength = sizeof(IPV6_HEADER) + sizeof(TCP_HEADER);
+        void *tcpResetPacket;
+        PIPV6_HEADER ipHeader;
+        PTCP_HEADER tcpHeader;
+
+        // allocate memory for the reset packet
+        tcpResetPacket = portmaster_malloc(packetLength, FALSE);
+
+        // initialize IPv6 header
+        ipHeader = (PIPV6_HEADER) tcpResetPacket;
+        ipHeader->Version = 6;
+        ipHeader->Length = sizeof(TCP_HEADER);
+        ipHeader->NextHdr = 6;
+        ipHeader->HopLimit = 128;
+        RtlCopyMemory(ipHeader->DstAddr, originalIPHeader->SrcAddr, sizeof(originalIPHeader->SrcAddr)); // Source becomes destination
+        RtlCopyMemory(ipHeader->SrcAddr, originalIPHeader->DstAddr, sizeof(originalIPHeader->DstAddr)); // Destination becomes source
+
+        // Initialize TCP header
+        tcpHeader = (PTCP_HEADER) ((UINT8*)tcpResetPacket + sizeof(IPV6_HEADER));
+        tcpHeader->SrcPort = RtlUshortByteSwap(packetInfo->remotePort); // Source becomes destination
+        tcpHeader->DstPort = RtlUshortByteSwap(packetInfo->localPort); // Destination becomes source
+        tcpHeader->HdrLength = sizeof(TCP_HEADER) / 4;
+        tcpHeader->SeqNum = 0;
+        // We should acknowledge the SYN packet while doing the reset
+        tcpHeader->AckNum = RtlUlongByteSwap(RtlUlongByteSwap(originalTCPHeader->SeqNum) + 1);
+        tcpHeader->Ack = 1;
+        tcpHeader->Rst = 1;
+
+        calc_ipv6_checksum(tcpResetPacket, packetLength, TRUE);
+
+        // Create the net buffer list for the new RST packet
+        status = wrap_packet_data_in_nb(tcpResetPacket, packetLength, &injectNBL);
+        if (!NT_SUCCESS(status)) {
+            ERR("send_tcp_rst ipv6 -> wrap_packet_data_in_nb failed: %u", status);
+            portmaster_free(tcpResetPacket);
+            return;
+        }
+
+        // loopback needs special treatment
+        if(is_ipv6_loopback(packetInfo->remoteIP)) {
```

Comment:
This *should* only be happening for outgoing packets, but we may want to block an incoming packet.
I suggest to check for loopback, or incoming packets.